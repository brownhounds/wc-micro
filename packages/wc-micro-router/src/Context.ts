import { EventBus } from '@brownhounds/event-bus';

import type {
    NavigateConfig,
    RouteObject,
    RouterConfig,
    RouterContextConfig,
    RouterEventsType,
} from './types';

export class RouterContext {
    public static config: RouterContextConfig = {
        routerTag: 'router-element',
        outletTag: 'outlet-element',
        routePrefix: '',
    };
    public static definedRoutes: RouteObject[];
    public static routes: RouteObject[] = [];
    public static eventBus = new EventBus<RouterEventsType, RouteObject>();
    public static navigate: (
        pathname: string,
        config?: NavigateConfig,
        popState?: boolean
    ) => void;
    public static navigateLocation: (pathname: string) => void;
    public static popState: (pathname: string) => void;

    public static mergeConfigs(config: RouterConfig): void {
        RouterContext.config = {
            ...RouterContext.config,
            ...config,
        };
    }

    public static handleError(error: Error): never {
        throw error;
    }
}
