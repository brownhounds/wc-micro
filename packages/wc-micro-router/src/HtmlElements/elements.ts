export * from './outlet-element';
export * from './router-element';
