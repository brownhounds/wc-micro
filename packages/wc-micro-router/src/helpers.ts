import { RouterContext } from './Context';
import type { RouteObject } from './types';

export const formatRoute = (route: string): string => {
    return (
        '/' +
        route
            .split('/')
            .filter((part) => part !== '')
            .join('/')
    );
};

const findDuplicatedValuesInArray = (array: string[]): string[] =>
    array.filter((e, i, a) => a.indexOf(e) !== i);

const findDuplicatedRoutes = (array: RouteObject[]): Set<string> => {
    const seen = new Set<string>();
    const duplicates = new Set<string>();

    for (const route of array) {
        if (seen.has(route.path)) duplicates.add(route.path);
        seen.add(route.path);
    }

    return duplicates;
};

const validateChildrenOutletIds = (routes: RouteObject[]): void => {
    const invalidRoute = routes.find((route) => !route.outletId);
    if (invalidRoute) {
        RouterContext.handleError(
            new Error(`Child Route: ${invalidRoute.path} has missing outletId`)
        );
    }
};

const validateUniqueRoutes = (): void => {
    const duplicates = findDuplicatedRoutes(RouterContext.routes);
    if (duplicates.size) {
        RouterContext.handleError(
            new Error(
                `Found duplicated routes: ${Array.from(duplicates).join(', ')}`
            )
        );
    }
};

const matchParamNames = (pattern: string): string[] =>
    pattern.match(/:[^/]+/g) || [];

const validateUniqueRouteParams = (): void => {
    for (const route of RouterContext.routes) {
        const duplicates = findDuplicatedValuesInArray(
            matchParamNames(route.path)
        );
        if (duplicates.length) {
            RouterContext.handleError(
                new Error(
                    `Route: ${route.path} has duplicated dynamic parameters ${duplicates.join(', ')}`
                )
            );
        }
    }
};

const matchParams = (
    pattern: string,
    pathname: string
): { paramNames: string[]; match: RegExpMatchArray } => {
    const regex = new RegExp(pattern.replace(/:[^/]+/g, '([^/]+)'));
    const match = pathname.match(regex);

    return {
        paramNames: matchParamNames(pattern),
        match: match as RegExpMatchArray,
    };
};

export const matchRoute = (pattern: string, pathname: string): boolean => {
    const regexPattern = pattern
        .replace(/:[a-zA-Z_]+/g, '([\\w-]+)')
        .replace(/\//g, '\\/');
    const regex = new RegExp(`^${regexPattern}$`);
    const match = pathname.match(regex);
    return match ? true : false;
};

export const castParamsToObject = (
    pattern: string,
    pathname: string
): Record<string, string> => {
    const { paramNames, match } = matchParams(pattern, pathname);
    const params: Record<string, string> = {};

    for (let i = 0; i < paramNames.length; i++) {
        const paramName = paramNames[i].slice(1);
        params[paramName] = match[i + 1];
    }

    return params;
};

const compileChildRoutes = (
    routes: RouteObject[],
    topLevelRoute: RouteObject,
    parentRoutePath: string = ''
): void => {
    for (const route of routes) {
        RouterContext.routes.push({
            ...route,
            path: formatRoute(
                RouterContext.config.routePrefix +
                    topLevelRoute.path +
                    parentRoutePath +
                    route.path
            ),
            topLevelRoute,
        });

        if (route.children) {
            validateChildrenOutletIds(route.children);
            compileChildRoutes(route.children, topLevelRoute, route.path);
        }
    }
};

export const compileRoutes = (): void => {
    for (const route of RouterContext.definedRoutes) {
        RouterContext.routes.push({
            ...route,
            path: formatRoute(RouterContext.config.routePrefix + route.path),
        });
        if (route.children?.length) {
            validateChildrenOutletIds(route.children);
            compileChildRoutes(route.children, route);
        }
    }

    validateUniqueRoutes();
    validateUniqueRouteParams();
};
