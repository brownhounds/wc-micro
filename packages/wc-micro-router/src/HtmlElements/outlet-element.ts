import { RouterContext } from '../Context';
import { matchRoute } from '../helpers';
import type { RouteObject } from '../types';
import { RouterEvents } from '../types';

export class OutletElement extends HTMLElement {
    #matchedRoute?: RouteObject;

    connectedCallback(): void {
        this.#ensureId();
        this.#matchedRoute = this.#matchRoute();
        this.#ensureRoute(this.#matchedRoute);
        RouterContext.eventBus.addListener(
            RouterEvents.NAVIGATE,
            this.#onNavigate
        );
        if (this.#matchedRoute)
            void this.#mountRouteElement(this.#matchedRoute);
    }

    disconnectedCallback(): void {
        RouterContext.eventBus.removeListener(
            RouterEvents.NAVIGATE,
            this.#onNavigate
        );
    }

    #onNavigate = (): void => {
        if (this.#matchedRoute)
            void this.#mountRouteElement(this.#matchedRoute);
    };

    #ensureId(): void {
        if (!this.id.length) {
            RouterContext.handleError(
                new Error(
                    `<${RouterContext.config.outletTag} /> element requires an id`
                )
            );
        }
    }

    #matchRoute(): RouteObject | undefined {
        return RouterContext.routes.find((route) => route.outletId === this.id);
    }

    #ensureRoute(route?: RouteObject): void | never {
        if (!route) {
            RouterContext.handleError(
                new Error(`Route with outletId: ${this.id} was not found`)
            );
        }
    }

    async #mountRouteElement(route: RouteObject): Promise<void> {
        const pathname = window.location.pathname
            .split('/')
            .splice(0, route.path.split('/').length)
            .join('/');

        if (matchRoute(route.path, pathname)) {
            if (this.firstElementChild?.localName !== route.component.tag) {
                await route.component.import();
                this.innerHTML = `<${route.component.tag}></${route.component.tag}>`;
            }
        } else {
            this.innerHTML = '';
        }
    }
}

if (customElements.get(RouterContext.config.outletTag) === undefined) {
    customElements.define(RouterContext.config.outletTag, OutletElement);
}
