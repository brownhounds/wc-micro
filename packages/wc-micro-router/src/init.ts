import { RouterContext } from './Context';
import { Router } from './Router';
import { compileRoutes } from './helpers';
import type { RouteObject, RouterConfig } from './types';

export const initRouter = (
    routes: RouteObject[],
    config?: RouterConfig
): Router => {
    if (config) RouterContext.mergeConfigs(config);
    RouterContext.definedRoutes = routes;

    compileRoutes();

    const router = new Router();
    RouterContext.navigate = router.navigate.bind(router);
    RouterContext.navigateLocation = router.navigateLocation.bind(router);
    RouterContext.popState = router.popState.bind(router);
    import('./HtmlElements/elements');

    return router;
};
