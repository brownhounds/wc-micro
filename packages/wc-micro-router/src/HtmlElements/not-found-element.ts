import { NOT_FOUND_TAG } from '../constants';

export class RouterNotFound extends HTMLElement {
    connectedCallback(): void {
        this.innerHTML = '<h1>404</h1><p>Page Not Found</p>';
    }
}

if (customElements.get(NOT_FOUND_TAG) === undefined) {
    customElements.define(NOT_FOUND_TAG, RouterNotFound);
}
