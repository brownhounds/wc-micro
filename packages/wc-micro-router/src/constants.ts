import type { ComponentDefinition } from './types';

export const NOT_FOUND_TAG = 'router-not-found';

export const builtInNotFound: ComponentDefinition = {
    tag: NOT_FOUND_TAG,
    import: () => import('./HtmlElements/not-found-element'),
};
