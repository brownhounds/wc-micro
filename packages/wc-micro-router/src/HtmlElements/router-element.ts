import { builtInNotFound } from '../constants';
import { RouterContext } from './../Context';
import type { RouteObject } from './../types';
import { RouterEvents } from './../types';

export class RouterElement extends HTMLElement {
    #onNavigate = async (payload?: RouteObject): Promise<void> => {
        if (payload) {
            const route = payload.topLevelRoute || payload;

            await route.component.import();

            if (this.firstElementChild?.localName !== route.component.tag)
                this.innerHTML = `<${route.component.tag}></${route.component.tag}>`;
            return;
        }

        void this.#onNotFound();
    };

    async #onNotFound(): Promise<void> {
        const element = RouterContext.config.notFound || builtInNotFound;
        if (this.firstElementChild?.localName !== element.tag) {
            await element.import();
            this.innerHTML = `<${element.tag}></${element.tag}>`;
        }
    }

    #onPopState = (e: PopStateEvent): void => {
        const target = e.target as Window;
        RouterContext.popState(target.location.pathname);
    };

    connectedCallback(): void {
        RouterContext.eventBus.addListener(
            RouterEvents.NAVIGATE,
            this.#onNavigate
        );
        RouterContext.navigateLocation(window.location.pathname);
        window.addEventListener('popstate', this.#onPopState);
    }

    disconnectedCallback(): void {
        RouterContext.eventBus.listeners.clear();
        window.removeEventListener('popstate', this.#onPopState);
    }
}

if (customElements.get(RouterContext.config.routerTag) === undefined) {
    customElements.define(RouterContext.config.routerTag, RouterElement);
}
