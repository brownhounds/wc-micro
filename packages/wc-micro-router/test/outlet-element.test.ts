import { expect, test, describe } from 'vitest';
import { OutletElement } from '../src/HtmlElements/outlet-element';
import { initRouter } from '../src/init';

describe('Route Matching', () => {
    test('Outlet', async () => {
        const OUTLET_ID = 'some-sort-of-id-here';
        initRouter([
            {
                path: '/',
                outletId: OUTLET_ID,
                component: {
                    tag: 'something-else',
                    import: (): Promise<unknown> => Promise.resolve(),
                },
            },
        ]);

        const outlet = new OutletElement();
        outlet.id = OUTLET_ID;
        expect(() => outlet.connectedCallback()).not.toThrow();
    });

    test('Outlet without id', async () => {
        const outlet = new OutletElement();
        expect(() => outlet.connectedCallback()).toThrowErrorMatchingSnapshot();
    });

    test('Outlet with id without a matching route', async () => {
        const outlet = new OutletElement();
        outlet.id = 'some-id';
        expect(() => outlet.connectedCallback()).toThrowErrorMatchingSnapshot();
    });
});
