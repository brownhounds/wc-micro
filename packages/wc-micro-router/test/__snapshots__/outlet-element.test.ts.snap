// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Route Matching > Outlet with id without a matching route 1`] = `[Error: Route with outletId: some-id was not found]`;

exports[`Route Matching > Outlet without id 1`] = `[Error: <outlet-element /> element requires an id]`;
